File: packages/kotlin/src/main/kotlin/com/wave/unlock/Gateway.kt
package com.wave.unlock

/** Config for the Wave gateway (Supabase edge functions). */
data class WaveConfig(
    val gatewayUrl: String,
    val anonKey: String,
    val publishableKey: String,
    val userNumber: String,
)

data class Outcome(val status: Status, val reason: String?) {
    enum class Status { GRANTED, DENIED, PENDING }
}

/** Gateway abstraction so the engine is testable without real HTTP. */
interface Gateway {
    suspend fun fetchToken(): String
    suspend fun readOutcome(token: String): Outcome
    suspend fun awaitOutcome(token: String, timeoutMs: Long, pollMs: Long = 500): Outcome
}

File: packages/flutter/lib/src/gateway.dart
/// Config for the Wave gateway.
class WaveConfig {
  final String gatewayUrl;
  final String anonKey;
  final String publishableKey;
  final String userNumber;
  const WaveConfig({
    required this.gatewayUrl,
    required this.anonKey,
    required this.publishableKey,
    required this.userNumber,
  });
}

enum OutcomeStatus { granted, denied, pending }

class Outcome {
  final OutcomeStatus status;
  final String? reason;
  const Outcome(this.status, this.reason);
}

/// Gateway abstraction so the engine is testable without real HTTP.
abstract class Gateway {
  Future<String> fetchToken();
  Future<Outcome> readOutcome(String token);
  Future<Outcome> awaitOutcome(String token, Duration timeout, {Duration poll});
}
